monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamResource.scala

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2014-2019 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.reactive.internal.builders
+
+import java.io.InputStream
+
+import monix.eval.Task
+import monix.execution.Ack.Continue
+import monix.execution.{Ack, AsyncVar, Cancelable, Scheduler}
+import monix.reactive.Observable
+import monix.reactive.observers.Subscriber
+
+import scala.annotation.tailrec
+import scala.concurrent.Await
+import scala.concurrent.duration.Duration
+
+class InputStreamResource(observable: Observable[Array[Byte]]) {
+
+  private type Bytes = Array[Byte]
+
+  def toTask: Task[InputStream] = {
+    Task.create[InputStream] { (s, cb) =>
+      val (state, cancelable) = subscribe(s)
+      cb.onSuccess(new ObservableInputStream(state, cancelable))
+    }
+  }
+
+  private def subscribe(s: Scheduler): (AsyncVar[Option[Bytes]], Cancelable) = {
+    val state = AsyncVar.empty[Option[Bytes]]()
+
+    val cancelable = observable.executeAsync.unsafeSubscribeFn(new Subscriber.Sync[Array[Byte]] {
+      implicit val scheduler = s
+
+      override def onNext(elem: Array[Byte]): Ack = {
+        state.put(Some(elem)).value
+        Continue
+      }
+
+      def onComplete(): Unit = {
+        state.put(None).value
+      }
+
+      def onError(ex: Throwable): Unit = {
+        state.put(None).value
+      }
+
+    })
+
+    (state, cancelable)
+  }
+
+  private[this] class ObservableInputStream(queue: AsyncVar[Option[Bytes]], cancelable: Cancelable)
+    extends InputStream {
+    private[this] var buffer: Array[Byte] = new Array[Byte](0)
+    private[this] var isClosed: Boolean = false
+
+    override def read(): Int = {
+      val a = new Array[Byte](1)
+      read(a, 0, 1) match {
+        case 1 => a(0) & 0xff
+        case -1 => -1
+        case len => throw new IllegalStateException(s"There was '1' byte expected, but there were '$len' bytes read")
+      }
+    }
+
+    override def read(arr: Array[Byte], begin: Int, length: Int): Int = {
+      require(arr.length > 0, "array size must be >= 0")
```

Comment:
This is something that I saw in the akka-streams code. Looks pretty logical, as this is sanity validation.
Instead of this validation I could:
- do nothing if array is empty
- do nothing if begin/length are negative 
- read only the amount of bytes that fit into the buffer.

However, as a user of the method I would prefer seeing exception, then having strange results in the byte array, that are caused by some internal implementation details.

Reply:
Makes sense!